docs/src/docs/asciidoc/user/serializers-copiers.adoc

```diff
@@ -0,0 +1,170 @@
+
+=== Copiers
+
+As the on-heap store is capable of storing plain java objects as such,
+it is not necessary to use a costly serialization and deserialization mechanism to copy keys and values when a
+on-heap store is configured to store by value. For instance, some form of "copy constructor" mechanism is a lot
+more performant but requires custom code to be able to copy objects.
+
+Copiers are specific to the on-heap store: no other store makes use of copiers.
+
+By default, the on-heap store stores by reference. The way to make it to store by value is to configure some copier(s)
+on it, either key, value or both.
+
+=== How is a serializer configured?
+
+There are two places where serializers can be configured:
+
+- at the cache level where one can use `CacheConfigurationBuilder.withKeySerializer()` and `CacheConfigurationBuilder.withValueSerializer()`.
+Both instances and class names can be configured.
+- at the cache manager level where one can use `CacheManagerBuilder.withSerializer(Class clazz, Class serializer)`
+
+If a serializer is configured directly at the cache level, it will be used.
+
+If a serializer is configured at the cache manager level, upon initialization, a cache with no specifically configured serializer will
+search through its cache manager's registered list of serializers and try to find one that directly matches the cache's key or value type. If such
+search fails, all the registered serializers will be tried in the added order to find one that handles compatible types.
+
+For instance, let's say you have a `Person` interface and two subclasses: `Employee` and `Customer`. If you configure your cache manager as follows:
+```
+CacheManagerBuilder.newCacheManagerBuilder().withSerializer(Employee.class, EmployeeSerializer.class).withSerializer(Person.class, PersonSerializer.class)
+```
+then configuring a `Cache<Long, Employee>` would make it use the `EmployeeSerializer` while a `Cache<Long, Customer>` would make it use the `PersonSerializer`.
+
+Each cache for which Ehcache instantiated a serializer (i.e.: the serializer was configured by its class name) will have its own instance.
+
+=== How is a copier configured?
+
+There is only one place where copiers can be configured: at the cache level where one can use `CacheConfigurationBuilder.withKeyCopier()`
+and `CacheConfigurationBuilder.withValueCopier()`.
+
+If a copier is configured at the cache level, that cache will work in store-by-value mode.
+
+Each cache for which Ehcache instantiated a copier (i.e.: the copier was configured by its class name) will have its own instance.
+
+=== Bundled implementations
+
+By default, cache managers are pre-configured with specially optimized serializers that can handle the following types, in the following order:
+
+- `java.io.Serializable`
+- `java.lang.Long`
+- `java.lang.Integer`
+- `java.lang.Float`
+- `java.lang.Double`
+- `java.lang.Character`
+- `java.lang.String`
```

Comment:
These are definitely not immutable types.

And I hope no one ever uses some as keys :wink: